src/clj/rems/db/duo.clj

```diff
@@ -0,0 +1,144 @@
+(s/defschema DuoCodeDb
+  {:id s/Str
+   (s/optional-key :shorthand) (s/maybe s/Str)
+   :label LocalizedString
+   :description LocalizedString})
+
+(def ^:private coerce-DuoCodeCsv
+  (coerce/coercer! DuoCodeCsv coerce/string-coercion-matcher))
+
+(def ^:private coerce-DuoCodeDb
+  (coerce/coercer! DuoCodeDb coerce/string-coercion-matcher))
+
+(def ^:private validate-DuoCodeDb
+  (s/validator DuoCodeDb))
+
+(defn- rows->maps
+  "Transform CSV `rows` into maps with the keys from the first header row."
+  [rows]
+  (mapv zipmap
+        (->> (first rows) ; header
+             (map keyword)
+             repeat)
+        (rest rows)))
+
+(defn upsert-duo-code!
+  "Insert a new or update an existing DUO code into the database."
+  [duo]
+  (let [data (-> duo
+                 validate-DuoCodeDb
+                 (dissoc :id)
+                 (->> (remove-vals #(cond (string? %) (str/blank? %)
+                                          (coll? %) (empty? %)
+                                          :else %)))
+                 json/generate-string)]
+    (when (:id duo)
+      (db/upsert-duo-code! {:id (:id duo)
+                            :data data}))))
+
+(defn- read-zip-entries
+  "Read the zip-file entries from the `stream` and returns those matching `re`."
+  [stream re]
+  (when stream
+    (loop [files []]
+      (if-let [entry (.getNextEntry stream)]
+        (if (re-matches re (.getName entry))
+          (let [baos (java.io.ByteArrayOutputStream.)]
+            (io/copy stream baos)
+            (recur (conj files {:name (.getName entry)
+                                :bytes (.toByteArray baos)})))
+          (recur files))
+        files))))
+
+(defn fetch-duo-releases-latest
+  "Fetches the latest DUO codes from the GitHub repository.
+
+  Returns the unpacked, formatted result."
+  []
+  (let [release-response (http/get "https://api.github.com/repos/EBISPOT/DUO/releases/latest"
+                                   {:accept :json
+                                    :as :json})
+        release-body (:body release-response)
+        release-tag (:tag_name release-body)
+        zip-response (http/get (:zipball_url release-body) {:as :stream})
+        csv-bytes (-> zip-response :body (java.util.zip.ZipInputStream.) (read-zip-entries #".*/duo.csv$") first :bytes)]
+    (with-open [reader (io/reader csv-bytes)]
+      (let [codes (for [duo (->> reader csv/read-csv rows->maps)]
+                    (-> duo
+                        coerce-DuoCodeCsv
+                        (update-existing :label (fn [en] {:en en}))
+                        (update-existing :description (fn [en] {:en en}))))]
+        {:tag release-tag
+         :codes codes}))))
+
+(def simple-codes
+  "Codes that can be added to a resource without any additional questions."
+  #{:DUO:0000004 :DUO:0000006 :DUO:0000011 :DUO:0000015 :DUO:0000016 :DUO:0000018 :DUO:0000019 :DUO:0000021 :DUO:0000029 :DUO:0000042 :DUO:0000043 :DUO:0000044 :DUO:0000045 :DUO:0000046})
```

Comment:
would it make sense to list these codes with the additional details as comments? like what the code is and so on

Reply:
That's something that is in the CSV file / DB so I wouldn't really want to duplicate it. On the other hand as the process is quite manual, we could have a helper function that generates a description of all the codes to be pasted into the source code. These were manually added by me based on what the description seems to say. They need to be manually checked every time one updates the codes (even though the promise is that they will not significantly change).

Reply:
i'm a little worried about the manual upkeeping. however if we keep the DUO codes in source code, i guess it would be more like a bonus on top to have the descriptions there. perhaps we can link to `http://purl.obolibrary.org/obo/`? e.g. `http://purl.obolibrary.org/obo/DUO_0000007`.

Reply:
As per the other question, we developers must follow up on the releases and update any codes / descriptions if they change. It would definitely be just a bonus. We can generate a commented version of the codes lists with e.g. the shorthand and label, and manually add any further details we want. I don't see a way of avoiding this.

Reply:
as continuation to this discussion, i looked up some of the codes in the OWL and i'm not 100% sure why some codes are listed in `simple-codes` and others in `complex-codes`.

example listed as `simple code`:
```
http://purl.obolibrary.org/obo/DUO_0000021

<obo:IAO_0000115 xml:lang="en">This data use modifier indicates that the requestor must provide documentation of local IRB/ERB approval.</obo:IAO_0000115>
```

example listed as `complex-code`:
```
http://purl.obolibrary.org/obo/DUO_0000022

<obo:IAO_0000115 xml:lang="en">This data use modifier indicates that use is limited to within a specific geographic region.</obo:IAO_0000115>
```

what would be the difference between these two from the systems standpoint?

also some codes are not in either listing, like `http://purl.obolibrary.org/obo/DUO_0000017` or `http://purl.obolibrary.org/obo/DUO_0000031` for example. are these obsolete codes?

Reply:
`DUO_0000017` was actually below with the comment `; not concrete types`

Reply:
Well my current understanding is such that you would need to ask some more info for each of the complex types i.e. disease, institute, attachment of document or so. We can get back to the details of what these turn out to be when we go through the details more. I think most of these can't be automatically checked for approval anyway.

In 22, there is effort to specify a geographic area data type etc. so I take it they want this to be asked. In 21 there simply isn't anything (in the OWL). Could be that there should be a question (the text seems like it) but since they didn't model it I simply left it as simple for now.

Reply:
I'm not sure if the not concrete types can be tagged or not. The data model kind of implies to me that one should use the more specific types always.